samples/bluetooth/broadcast_audio_sink/src/main.c

```diff
@@ -76,6 +78,97 @@ static uint32_t requested_bis_sync;
 static uint32_t bis_index_bitfield;
 static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
 
+#if defined(CONFIG_LIBLC3)
+
+#include "lc3.h"
+
+#define MAX_SAMPLE_RATE       16000
+#define MAX_FRAME_DURATION_US 10000
+#define MAX_NUM_SAMPLES       ((MAX_FRAME_DURATION_US * MAX_SAMPLE_RATE) / USEC_PER_SEC)
+
+static int16_t audio_buf[MAX_NUM_SAMPLES];
+static lc3_decoder_t lc3_decoder;
+static lc3_decoder_mem_16k_t lc3_decoder_mem;
+static int frames_per_sdu;
+
+static int lc3_enable(const struct bt_audio_codec_cfg *codec_cfg)
+{
+	int8_t ret;
+	int freq_hz;
+	int frame_duration_us;
+
+	printk("Enable: stream with codec %p\n", codec_cfg);
+
+	ret = bt_audio_codec_cfg_get_freq(codec_cfg);
+	if (ret > 0) {
+		freq_hz = bt_audio_codec_cfg_freq_to_freq_hz(ret);
+	} else {
+		printk("Error: Codec frequency not set, cannot start codec.");
+		return -1;
+	}
+
+	ret = bt_audio_codec_cfg_get_frame_dur(codec_cfg);
+	if (ret > 0) {
+		frame_duration_us = bt_audio_codec_cfg_frame_dur_to_frame_dur_us(ret);
+	} else {
+		printk("Error: Frame duration not set, cannot start codec.");
+		return ret;
+	}
+
+	frames_per_sdu = bt_audio_codec_cfg_get_frame_blocks_per_sdu(codec_cfg, true);
+
+	lc3_decoder = lc3_setup_decoder(frame_duration_us, freq_hz, 0, /* No resampling */
+					&lc3_decoder_mem);
+
+	if (lc3_decoder == NULL) {
+		printk("ERROR: Failed to setup LC3 decoder - wrong parameters?\n");
+		return -1;
+	}
+	return 0;
+}
+
+static void lc3_decode_handler(struct k_work *work)
+{
+	int err = 0;
+	int offset = 0;
+	struct broadcast_sink_stream *sink_stream = CONTAINER_OF(
+		k_work_delayable_from_work(work), struct broadcast_sink_stream, lc3_decode_work);
+
+	err = k_sem_take(&sem_bis_lc3_decode, K_NO_WAIT);
```

Comment:
You only ever take `sem_bis_lc3_decode`, but you don't release it. 

It also needs to taken in `stream_recv_cb`, so that you don't ever have 2 threads accessing `sink_stream->in_buf` at the same time.